google-cloud-firestore/src/main/java/com/google/cloud/firestore/Order.java

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright 2017 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.firestore;
+
+import com.google.firestore.v1beta1.Value;
+import com.google.firestore.v1beta1.Value.ValueTypeCase;
+import com.google.protobuf.ByteString;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map.Entry;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import javax.annotation.Nonnull;
+
+/** Order implements the ordering semantics of the backend. */
+class Order implements Comparator<Value> {
+  enum TypeOrder implements Comparable<TypeOrder> {
+    // NOTE: This order is defined by the backend and cannot be changed.
+    NULL,
+    BOOLEAN,
+    NUMBER,
+    TIMESTAMP,
+    STRING,
+    BLOB,
+    REF,
+    GEO_POINT,
+    ARRAY,
+    OBJECT;
+
+    static TypeOrder fromValue(Value value) {
+      switch (value.getValueTypeCase()) {
+        case NULL_VALUE:
+          return NULL;
+        case BOOLEAN_VALUE:
+          return BOOLEAN;
+        case INTEGER_VALUE:
+          return NUMBER;
+        case DOUBLE_VALUE:
+          return NUMBER;
+        case TIMESTAMP_VALUE:
+          return TIMESTAMP;
+        case STRING_VALUE:
+          return STRING;
+        case BYTES_VALUE:
+          return BLOB;
+        case REFERENCE_VALUE:
+          return REF;
+        case GEO_POINT_VALUE:
+          return GEO_POINT;
+        case ARRAY_VALUE:
+          return ARRAY;
+        case MAP_VALUE:
+          return OBJECT;
+        default:
+          throw new IllegalStateException("Could not detect value type for " + value);
+      }
+    }
+  }
+
+  /**
+   * Main comparison function for all Firestore types.
+   *
+   * @return -1 is left < right, 0 if left == right, otherwise 1
+   */
+  public int compare(@Nonnull Value left, @Nonnull Value right) {
+    // First compare the types.
+    TypeOrder leftType = TypeOrder.fromValue(left);
+    TypeOrder rightType = TypeOrder.fromValue(right);
+    int typeComparison = leftType.compareTo(rightType);
+    if (typeComparison != 0) {
+      return typeComparison < 0 ? -1 : 1;
+    }
```

Comment:
There are lots of these "double compares". Do we care that deeply that the returned values are only -1,0,or 1?

Reply:
I rely on this in the test case. But it makes much more sense to convert the end result in the test case itself, so I moved that conversion there.

Reply:
I moved this conversion to the test class.